compat/openssl.h

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+#ifndef OPENSSL_COMPAT_H
+#define OPENSSL_COMPAT_H
+
+#if defined(__cplusplus)
+
+#if HAVE_OPENSSL_ASN1_H
+#include <openssl/asn1.h>
+#endif
+#if HAVE_OPENSSL_BIO_H
+#include <openssl/bio.h>
+#endif
+#if HAVE_OPENSSL_DH_H
+#include <openssl/dh.h>
+#endif
+#if HAVE_OPENSSL_EVP_H
+#include <openssl/evp.h>
+#endif
+#if HAVE_OPENSSL_LHASH_H
+#include <openssl/lhash.h>
+#endif
+#if HAVE_OPENSSL_SSL_H
+#include <openssl/ssl.h>
+#endif
+#if HAVE_OPENSSL_X509_H
+#include <openssl/x509.h>
+#endif
+
+
+#if !HAVE_LIBCRYPTO_ASN1_STRING_GET0_DATA
+inline const unsigned char *
+ASN1_STRING_get0_data(const ASN1_STRING *x)
+{
+    return x->data;
+}
+#endif
+
+#if !HAVE_LIBCRYPTO_BIO_GET_DATA
+inline void *
+BIO_get_data(BIO *table)
+{
+    return table->ptr;
+}
+
+inline void
+BIO_set_data(BIO *table, void *data)
+{
+   table->ptr = data;
+}
+
+inline void
+BIO_set_init(BIO *table, int init)
+{
+   table->init = init;
+}
+#endif
+
+// BIO_get_init is not implemented in LibreSSL, even though BIO_set_init is
+#if !HAVE_LIBCRYPTO_BIO_GET_INIT
+inline int
+BIO_get_init(BIO *table)
+{
+    return table->init;
+}
+#endif
+
+#if !HAVE_LIBCRYPTO_DH_UP_REF // OpenSSL 1.1 API
+#if defined(CRYPTO_LOCK_DH) // OpenSSL 1.0 API
+inline int
+DH_up_ref(DH *t)
+{
+    if (t)
+        CRYPTO_add(&t->references, 1, CRYPTO_LOCK_DH);
+    return 0;
+}
+#else
+#error missing both OpenSSL API features DH_up_ref (v1.1) and CRYPTO_LOCK_DH (v1.0)
+#endif /* OpenSSL 1.0 CRYPTO_LOCK_DH */
+#endif /* OpenSSL 1.1 DH_up_ref */
+
+#if !HAVE_LIBCRYPTO_EVP_PKEY_GET0_RSA
+inline RSA *
+EVP_PKEY_get0_RSA(EVP_PKEY *pkey)
+{
+    if (pkey->type != EVP_PKEY_RSA)
+        return nullptr;
+    return pkey->pkey.rsa;
+}
+#endif
+
+#if !HAVE_LIBCRYPTO_EVP_PKEY_UP_REF
+#if defined(CRYPTO_LOCK_EVP_PKEY) // OpenSSL 1.0
+inline int
+EVP_PKEY_up_ref(EVP_PKEY *t)
+{
+    if (t)
+        CRYPTO_add(&t->references, 1, CRYPTO_LOCK_EVP_PKEY);
```

Comment:
This function is not implemented correctly.  I can't return `CRYPTO_add` directly, since it returns the updated reference count, while `EVP_PKEY_up_ref` should return 1 for success and 0 for failure.  As a note, we were always signaling failure; apparently, the return values are never checked.  I will fix all of the similar `.*_up_ref` functions that have this same behaviour, using the same logic as openssl > 1.1 does, returning 1 if the updated reference count > 1, and 0 otherwise.